diff --git a/pylon_camera/config/default.yaml b/pylon_camera/config/default.yaml
index 5d990a4..bc3dbba 100644
--- a/pylon_camera/config/default.yaml
+++ b/pylon_camera/config/default.yaml
@@ -36,7 +36,7 @@ frame_rate: 5.0
 trigger_timeout: 5000  # in ms
 grab_timeout: 500  # in ms
 
-# Camera grab strategy 
+# Camera grab strategy
 # 0 : GrabStrategy_OneByOne
 # 1 : GrabStrategy_LatestImageOnly
 # 2 : GrabStrategy_LatestImages
@@ -44,7 +44,7 @@ grab_strategy : 0
 
 ##--White Balance Auto--##
 # Automatically corrects color shifts in images acquired. 0 = Off, 1 = Once, 2 = Continuous
-# white_balance_auto : 0 
+# white_balance_auto : 0
 ##----------------------##
 
 ##--White balance--##
diff --git a/pylon_camera/launch/pylon_camera_node.launch b/pylon_camera/launch/pylon_camera_node.launch
index 34c322c..5e3ab8b 100644
--- a/pylon_camera/launch/pylon_camera_node.launch
+++ b/pylon_camera/launch/pylon_camera_node.launch
@@ -1,5 +1,6 @@
 <?xml version="1.0"?>
 <launch>
+    <arg name="calibration" default="$(find pylon_camera)/../../../calibration" />
     <arg name="respawn" default="false" />
     <arg name="debug" default="false" />
     <arg name="node_name" default="pylon_camera_node" />
@@ -19,6 +20,8 @@
         <param name="startup_user_set" value="$(arg startup_user_set)"/>
         <param name="enable_status_publisher" value="$(arg enable_status_publisher)"/>
         <param name="enable_current_params_publisher" value="$(arg enable_current_params_publisher)"/>
+		<param name="camera_info_url" value="file://$(arg calibration)/ost.yaml"/>
+
     </node>
 
 </launch>
